specs/Standards/MultiToken/Events.md

```diff
@@ -0,0 +1,271 @@
+Events are log entries that start with `EVENT_JSON:` prefix followed by a single valid JSON document of the following interface:
+
+```ts
+// Interface to capture data
+// about an event
+// Arguments
+// * `standard`: name of standard e.g. nep171
+// * `version`: e.g. 1.0.0
+// * `event`: string
+// * `data`: associate event data
+interface EventLogData {
+    standard: string,
+    version: string,
+    event: string,
+    data?: unknown,
+}
+```
+
+#### Valid event logs:
+
+```js
+EVENT_JSON:{"standard": "nepXXX", "version": "1.0.0", "event": "xyz_is_triggered"}
+```
+
+```js
+EVENT_JSON:{
+  "standard": "nepXXX",
+  "version": "1.0.0",
+  "event": "xyz_is_triggered"
+}
+```
+
+```js
+EVENT_JSON:{"standard": "nepXXX", "version": "1.0.0", "event": "xyz_is_triggered", "data": {"triggered_by": "foundation.near"}}
+```
+
+#### Invalid event logs:
+
+* Two events in a single log entry (instead, call `log` for each individual event)
+```
+EVENT_JSON:{"standard": "nepXXX", "version": "1.0.0", "event": "xyz_is_triggered"}
+EVENT_JSON:{"standard": "nepXXX", "version": "1.0.0", "event": "xyz_is_triggered"}
+```
+* Invalid JSON data
+```
+EVENT_JSON:invalid json
+```
+
+
+## Interface
+Multi Token Events MUST have `standard` set to `"nep246"`, standard version set to `"1.0.0"`, `event` value is one of `mt_mint`, `mt_burn`, `mt_transfer`, `mt_approval` and `data` must be of one of the following relavant types: `MtMintLog[] | MtBurnLog[] | MtTransferLog[] | MtApprovalLog[]`:
+
+
+
+```ts
+// Interface to capture an event. The indexer looks for events prefixed 
+// with EVENT_JSON. The rest follows the EventLog 
+type MtEvent = "mt_mint" | "mt_burn" | "mt_transfer" | "mt_approval"
+
+// Interface for MT contract event data. It is used to emit event data with
+// the near standard logging.    
+// * `EVENT_JSON`: The standard event prefix required to signal consumers about
+// the type of log data being emitted.
+// * `standard`: name of standard e.g. nep-246 
+// * `version`: e.g. "1.0.0"
+// * `event`: `mt_mint` | `mt_burn` | `mt_transfer` | `mt_approval`
+// * `data`: associate event data
+interface MtEventLogData {
+  EVENT_JSON: {
+    standard: "nep246",
+    version: "1.0.0",
+    event: MtEvent,
+    data: MtMintLog[] | MtBurnLog[] | MtTransferLog[] | MtApprovalLog[]
+  }
+}
+
+// Minting event log. Emitted when a token is minted/created. 
+// Requirements
+// * Contract MUST emit event when minting a token
+// Fields 
+// * Contract token_ids and amounts MUST be the same length 
+// * `owner_id`: the account receiving the minted token
+// * `token_ids`: the tokens minted
+// * `amounts`: the number of tokens minted, wrapped in quotes and treated
+//    like a string, although the numbers will be stored as an unsigned integer
+//    array with 128 bits.
+// * `memo`: optional message
+interface MtMintLog {
+    owner_id: string,
+    token_ids: string[],
+    amounts: string[],
+    memo?: string
```

Comment:
I am sorry if I'm telling stupid things, it's hard to process all this amount of data.

Do we need Events section here?

As I understand, inside MultiToken contract, we still have FT or NFT (or both). Even if we are thinking about semi-fungible token, it is still fungible token at first, non-fungible token after.

In the events level of abstraction, we don't care about the concept of Multi Token. The contract creators should provide logs about FTs or NFTs. From the Indexer side, it's OK if one contract will provide both FT and NFT logs (it could be even done inside one receipt, though they should be different log statements: it's not a problem at all).
Indeed, it's better to mention the details about semi-fungible token explicitly:
1. Creating semi-fungible token is the mint of FT
2. Converting semi-fungible token from FT to NFT should be done by burning FT and minting NFT (I understand it this way).
3. You could transfer the tokens at any time by regular rules of transferring FT/NFT

When we will think about FT events (BTW, please have a look at [my initial thoughts](https://github.com/near/NEPs/issues/271) about it), we have to add "token_id" column, I will add that to requirements. That's the only way to distinguish between FTs inside one MultiToken contract.

Reply:
Hey great questions. 

I think we do need an events section. I think the argument for the events section is that the MT standard is just a different type of token. The standard by design is meant to be a more flexible standard to support alot of varied and new use cases. The background here is that the standard isn't really supposed to explicitly contain FT or NFTs, just have the means necessary to represent them within the token.  Use case wise the options you listed make sense, probably to add to that list are things like transferring multiple fungible tokens for NFTs, and vice versa.


From a background processing task perspective:

If you are consuming FT or NFT events, you might not understand from a consuming point of view what type of token standard you're operating on. I think you lose the larger context that might be important ( what contract am I operating on). 

Let's say you want to run a market place and respond to being approved for a transfer, the event would come over the wire and you wouldn't know what to call mt_transfer or ft_transfer/nft_transfer .

Let's say a transfer just occurred and you wanted, react to that transfer by using the rainbow bridge to relay that tx, how do you know what contract type to line up. Hypothetical situation and I might be off base with this example, but the issue is I think context :).  

Maybe additionally a bit of separation of concerns, i.e. multi token standard can evolve as it needs without worrying about ft or nft standard changes, and vice versa.  

Does that make sense?

Reply:
@zcstarr I feel there is some miscommunication here, so let me put my 2 cents here.

> I think we do need an events section.

I agree that this standard has to have an explicit events section, but ...

> I think the argument for the events section is that the MT standard is just a different type of token.

This seems not accurate to me. Multi-token is an umbrella standard based upon the prior art.

> The standard by design is meant to be a more flexible standard to support alot of varied and new use cases. 

The more you add to a standard, the fewer chances you have for contracts implementing all of that, and thus it defeats the purpose of having a standard in the first place.

The more optional features you have, the more fragmented your ecosystem becomes, and thus it defeats the purpose of having a standard in the first place.

> If you are consuming FT or NFT events, you might not understand from a consuming point of view what type of token standard you're operating on.

We can easily distinguish the types of events because each event log has to have `standard` field: https://nomicon.io/Standards/NonFungibleToken/Event.html#examples

Thus, if MT contract emits:

```js
EVENT_JSON:{
  "standard": "nep171",
  "version": "1.0.0",
  "event": "nft_mint",
  "data": [
    {"owner_id": "foundation.near", "token_ids": ["aurora", "proximitylabs"]}
  ]
}
EVENT_JSON:{
  "standard": "nep141", // NOTE that it is 141, not 171 as it is above
  "version": "1.0.0",
  "event": "ft_mint",
  "data": [
    {"owner_id": "foundation.near", /* yet to be defined in FT standard */}
  ]
}
```

Having MT-specific events only makes sense for those events that are defined in neither FT nor NFT standards. If we re-use the existing event standards, we achieve better composability and external tooling will just work out of the box.

> Let's say you want to run a market place and respond to being approved for a transfer, the event would come over the wire and you wouldn't know what to call mt_transfer or ft_transfer/nft_transfer .

One should just query contract metadata to check what interfaces the contract support. Also, it might be a good idea to support `nft_transfer*` and `ft_transfer*` methods in MT standard to make the interface composable.

> multi token standard can evolve as it needs without worrying about ft or nft standard changes, and vice versa.

I want to make sure we are on the same page, *standards cannot evolve in a non-compatible way*. Any breaking change will split the ecosystem. Thus, I suggest we don't split it in the first place and have composability with FT / NFT from the get go.

Reply:
This is great :). These are good points to bring up. 

> This seems not accurate to me. Multi-token is an umbrella standard based upon the prior art.


Let me provide a bit of context. At the beginning we spent some time discussing what is the standard . https://github.com/shipsgold/multi-token-standard-impl/blob/main/meetings/2021-07-07-meeting-notes.md#quick-summary

The takeaway was that we wanted an erc-1155 equivalent standard for Near. As we went along the journey, we had a much thinner standard, one that didn't define metadata and left it all to URI (erc-1155 equivalence).

After some discussion with community members, wallet team etc... . We came to a new conclusion which was that we should just support all metadata from the existing FT and NFT standards, on chain because we could, without major negative gas implications ( as on Ethereum). 

The erc-1155 standard doesn't have things like symbol or decimals, supply or any other associated FT/NFT metadata, mainly because things like video games don't need them and they could take up space for other use cases. The thought with 1155, was that people would use events aggregation to figure out state level things like supply etc... . We're not doing that but it's just the historical context :D .

Overall  premise is with a token_id, you can map that to whatever definition of fungibility you'd like. The only real key is the token_id and what you make of it. So it's not really directly tied to FT or NFT. In practice it tends to be, but the idea is to be lighter weight and less restrictive on what the token could be and keep as much equivalence as possible with 1155 for bridging.

>The more you add to a standard, the fewer chances you have for contracts implementing all of that, and thus it defeats the purpose of having a standard in the first place.
>The more optional features you have, the more fragmented your ecosystem becomes, and thus it defeats the purpose of having a standard in the first place.

Totally agree that a standard should be as thin as possible and as rigid as necessary.
>Thus, if MT contract emits:
```
EVENT_JSON:{
  "standard": "nep246", //NOTE that it is 246, not 171
  "version": "1.0.0",
  "event": "nft_mint",
  "data": [
    {"owner_id": "foundation.near", "token_ids": ["aurora", "proximitylabs"]}
  ]
}
EVENT_JSON:{
  "standard": "nep246", // NOTE that it is 246, not 141 as it is above
  "version": "1.0.0",
  "event": "ft_mint",
  "data": [
    {"owner_id": "foundation.near", /* yet to be defined in FT standard */}
  ]
}
```
Is the suggestion here that we just change the standard that's being emitted?  

If so I'm on board with that . My concern was about context for the events. I forgot that we could change the event standard field. So I'm all good with having that be emitted same as the other tokens. As long as we change the standard field.

>I want to make sure we are on the same page, standards cannot evolve in a non-compatible way. Any breaking change will split the ecosystem. Thus, I suggest we don't split it in the first place and have composability with FT / NFT from the get go.

Generally on the same page. My comment was about having a system, where say the MT standard needed to add have an additional field or interface for some unforseen purpose. There's no need to go and touch all the FT and NFT standards to also have this field/interface that's not relevant to those standards. The MT standard could evolve as it needs to without it needing to add interfaces in all the standards, because it's still loosely coupled. So non breaking changes.

Reply:
For clarity and conciseness here T_T. I'm good with events being shared across the standard and changing the standard field of the events. If we have consensus on this then I'll make that change. The approval thing is a nice to have maybe it can be for a later date or something all the standards can utilize.

Reply:
@frol @telezhnaya  How would you like me to proceed 

The thought would be to refactor Events.md to just have references to the NFT minting and then the FT minting/burning. It's unclear what the next steps should be there. 

Option A)  Yank events from the standard for now. 
Option B) Create a PR to land a FT Events Standard. with optional token_id path. Once the FT Events Standard lands. Refactor Multi Token standard Events.md to refer to the FT and NFT Events Standards.
Option C) Something else ?

Reply:
I've read https://eips.ethereum.org/EIPS/eip-1155 more carefully

@zcstarr said earlier
> I think the argument for the events section is that the MT standard is just a different type of token.

I need to reply on it, this is not true at all. EIP1155 is an umbrella standard, it does not create the new type of token. 
If we want to support semi-fungible tokens (or anything else) in our standard, we need to create the separate standard for semi-fungible token, and then add the support of it to this umbrella standard.

I do not understand the idea of the token with separate MT type. If I want to use NFT, I want NFT to be created, same with FT.

> The only real key is the token_id and what you make of it. So it's not really directly tied to FT or NFT. 

I disagree with it, the type is extremely important, it tells us the context how it could be used.

> [...] the idea is to [...] keep as much equivalence as possible with 1155 for bridging.

It means that we don't want to introduce new Multi-Token type.

> How would you like me to proceed

It's better to land FT events at first, but I'm not sure whether it's an option.
I think we should explain that MT contracts operate with FT/NFT. It means that MT contract should provide event logs as it's described in FT/NFT standard. So, here, in the `events.md`, we need just give general explanations and the links to `events.md` for all supported token types. 

@zcstarr @mikedotexe @frol what do you think?